### ISSUE TITLE
Nuke plugin build failure on Mac

### ISSUE THREAD

**Issue by danherman** [OpenImageIO/oiio#1301]
I'm seeing a problem superficially similar to https://github.com/OpenImageIO/oiio/issues/1155, but I have that fix and my `USE_NUKE` is 1.

My `NUKE_HOME` is `/Applications/Nuke9.0v8/Nuke9.0v8.app/Contents/MacOS/`

Build reports:

> Scanning dependencies of target txReader
> [ 98%] Building CXX object src/nuke/txReader/CMakeFiles/txReader.dir/txReader.cpp.o
> [ 99%] Linking CXX shared library txReader.dylib
> clang: error: invalid argument '-bundle' not allowed with '-dynamiclib'
> make[3]: **\* [src/nuke/txReader/txReader.dylib] Error 1
> make[2]: **\* [src/nuke/txReader/CMakeFiles/txReader.dir/all] Error 2
> make[1]: **\* [all] Error 2
> make: **\* [cmake] Error 2

Any ideas where to go with this? Thanks for ideas.

**Comment by lgritz** [comment]
I don't have Nuke on my Mac, so I can't test this easily myself.

Shot in the dark: what happens if in src/nuke/txReader/CMakeLists.txt (and the one in src/nuke/txWriter), you just remove the "-bundle" from the LINK_FLAGS definition?

Does it really need to be a bundle? Or just a regular dso?

To be honest, despite using Mac's for years, I tend to use them as just another Unix/Linux box (but with better compatibility with apps and devices), and have no idea what distinguishes a bundle from a shared library. ELI5?

**Comment by lgritz** [comment]
Also, you may want to ask questions like this on the oiio-dev mail list. It casts a much wider net than filing an "issue", which will be seen only by the relatively few people who "watch" the GitHub project.

"Issues" tend to work well when you have isolated a bug and we want to track it and know when it's solved, but are not well-suited to open ended "how do I" kinds of questions.

**Comment by danherman** [comment]
Sorry for the slow response... holidays took over. Removing "-bundle" produced a new error:

Scanning dependencies of target txReader
[ 98%] Building CXX object src/nuke/txReader/CMakeFiles/txReader.dir/txReader.cpp.o
[ 99%] Linking CXX shared library txReader.dylib
Undefined symbols for architecture x86_64:
  "DD::Image::MetaData::Bundle::setData(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*)", referenced from:
      TxReaderNS::txReader::fillMetadata(OpenImageIO::v1_7::ImageSpec const&, bool) in txReader.cpp.o
  "DD::Image::MetaData::Bundle::setData(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      TxReaderNS::txReader::fillMetadata(OpenImageIO::v1_7::ImageSpec const&, bool) in txReader.cpp.o
  "DD::Image::MetaData::Bundle::setData(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, float)", referenced from:
      TxReaderNS::txReader::txReader(DD::Image::Read*) in txReader.cpp.o
      TxReaderNS::txReader::fillMetadata(OpenImageIO::v1_7::ImageSpec const&, bool) in txReader.cpp.o
  "DD::Image::MetaData::Bundle::setData(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int)", referenced from:
      TxReaderNS::txReader::txReader(DD::Image::Read*) in txReader.cpp.o
      TxReaderNS::txReader::fillMetadata(OpenImageIO::v1_7::ImageSpec const&, bool) in txReader.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [src/nuke/txReader/txReader.dylib] Error 1
make[2]: *** [src/nuke/txReader/CMakeFiles/txReader.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [cmake] Error 2

I'm going to leave this there for the time being and retreat to Linux. Sorry, didn't know about the oiio-dev mail list, thanks for pointing it out.

**Comment by nrusch** [comment]
Unfortunately, I didn't (and still don't) have an OSX box with GCC 4.0 and the 10.6 SDK available when I was writing those CMake files. I basically just tried to copy the compiler and linker flags as closely as possible from the sample Makefile that ships with Nuke, but it looks like some new flags were added for Nuke 9, or maybe I just screwed something up.

If you want to take a stab at modifying the CMake files for those plugins, you can find the example Makefile in (e.g.) `/Applications/Nuke9.0v7/Documentation/NDK/examples`. From taking a quick look, I would try changing the relevant block to this:

```cmake
elseif (APPLE)
    target_link_libraries (txReader
        GLEW)
    add_definitions (-DUSE_GLEW)
    set_target_properties (txReader PROPERTIES
        PREFIX ""
        COMPILE_FLAGS "-isysroot /Developer/SDKs/MacOSX10.6.sdk -arch x86_64"
        LINK_FLAGS "-Wl,-syslibroot,/Developer/SDKs/MacOSX10.6.sdk -arch x86_64 -bundle -framework QuartzCore -framework IOKit -framework CoreFoundation -framework Carbon -framework ApplicationServices -framework OpenGL -framework AGL")
```

You may need to modify the paths to the 10.6 SDK based on your installation path. You may also need to add a `-F` compile flag that points to `Nuke9.0v7.app/Contents/Frameworks`.

**Comment by lgritz** [comment]
Should I add this? I have no way to test it myself, but if you guys say that what's there now is certainly broken and this can't make it any worse (and might fix it), I'm ok with speculatively merging it.

**Comment by nrusch** [comment]
The sysroot build flags are definitely required as far as I can tell, but I don't know if it's worth committing that change alone without knowing for sure if things are going to work. If you feel it's not a big deal, though, feel free. I'm going to see if I can find a box stashed somewhere around the office that still has the necessary build tools installed, and if I can, I'll try to get this figured out and submit a proper PR. 

@danherman: I don't know if you've taken a look at [this page](https://www.thefoundry.co.uk/products/nuke/developers/90/ndkdevguide/appendixa/osx.html) already, but the Foundry's build requirements are pretty specific (and old).

**Comment by melMass** [comment]
Hello Guys,

We are in 2018 and I was wondering if there is updates about this issue ?
The foundry did [updated their build system lately](https://learn.foundry.com/nuke/developers/110/ndkdevguide/appendixa/osx.html), so no need to install older unmaintained Snow Leopard.

**Comment by lgritz** [comment]
I don't really have a good way to test this myself. I reiterate again that if somebody submits a proper PR that addresses this issue, I will happily merge this.

**Comment by lgritz** [comment]
This really needs somebody with the Mac + Nuke combination to test out the proposed solution and submit a proper PR. I'll leave this open a little while more, but next time I come across this stale ticket, I will close if nobody has stepped forward either with a PR or at least to say that this is an ongoing issue for them.

**Comment by lgritz** [comment]
No further comments that this is a problem --> closing. Please feel free to re-open or submit a fix if this is a continued problem. I don't have a way to reproduce it or test a fix myself.
